### PR TITLE
new picai seg

### PR DESCRIPTION
**Pull request by itaijj**

### CODE REVIEW

File: fuse_examples/tests/conf/config.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  train_folds : [0,1,2,3]
+  validation_folds : [4]
+  batch_size: 16
+  learning_rate : 1e-3
+  weight_decay :  0
+  resume_checkpoint_filename : 
+  trainer:
+    accelerator : gpu
+    devices : 1
+    num_epochs : 100
```
Comment (alex-golts) [inline fuse_examples/tests/conf/config.yaml:30]: you do 100 epochs just for a unit test?

File: fuse_examples/tests/test_segmentation_picai.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+def run_picai(root: str) -> None:
+    initialize(config_path="conf", job_name="test_app")
+    cfg = compose(config_name="config")
+    print(str(cfg))
+
+    cfg = NDict(hydra.utils.instantiate(cfg))
+    cfg["paths.working_dir"] = root
+    cfg["train.run_sample"] = 100
+    cfg["train.trainer.num_epochs"] = 1
```
Comment (alex-golts) [inline fuse_examples/tests/test_segmentation_picai.py:49]: oh you modify it here

File: fuseimg/datasets/picai.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            for seq in self._sequences:
+                img_filename = os.path.join(
+                    self._dir_path, sample_dict[key_in].split("_")[0], sample_dict[key_in] + seq + ".mha"
+                )
+
+                image_data, image_header = load(img_filename)
+                sample_dict[key_out + seq] = image_data
+            return sample_dict
+        except:
+            return None
```
Comment (alex-golts) [inline fuseimg/datasets/picai.py:74]: maybe you should display a clearer warning when this happens?

File: fuseimg/datasets/picai.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+class OpLoadPICAISegmentation(OpBase):
+    """
+    Loads a medical image
+    """
+
+    def __init__(self, data_dir: str, dir_path: str, seuqences: Sequence[str] = ["_t2w"], **kwargs: dict):
+        super().__init__(**kwargs)
+        self._dir_path = dir_path
+        self._data_dir = data_dir
+        # self._sequences = seuqences
```
Comment (alex-golts) [inline fuseimg/datasets/picai.py:86]: remove comment

**Review by alex-golts — DISMISSED** [review]
LGTM
some small comments inline

File: fuse_examples/imaging/segmentation/picai/conf/config_template.yaml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    apply_aug_prob: 0.9
+    rotate: [-180, 180]
+    scale: [0.9, 2]
+    flip: 0.5
+    translate: [-25, 25]
+    mask_prob: 0
+    total_mask_prob: 0
+  loss_config :
+    segmentation_lambda : 0.2
+    classification_lamba : 0.8
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/conf/config_template.yaml:50]: those params will be removed, it is from old version actually

File: fuse_examples/imaging/segmentation/picai/runner.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+from torch.utils.data.dataloader import DataLoader
+
+from fuse.utils.utils_logger import fuse_logger_start
+from fuse.utils import NDict
+from fuse.data.utils.samplers import BatchSamplerDefault
+from fuse.data.utils.collates import CollateDefault
+from fuse.data.utils.split import dataset_balanced_division_to_folds
+from fuse.dl.losses.loss_default import LossDefault
+
+# from report_guided_annotation import extract_lesion_candidates
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/runner.py:35]: done

File: fuse_examples/imaging/segmentation/picai/runner.py [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+#     neural_network_for_run
+import torch
+import hydra
+from omegaconf import DictConfig, OmegaConf
+
+# from monai.losses import DiceFocalLoss
+
+# assert (
+#     "PICAI_DATA_PATH" in os.environ
+# ), "Expecting environment variable CMMD_DATA_PATH to be set. Follow the instruction in example README file to download and set the path to the data"
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/runner.py:59]: fixed

File: fuse_examples/imaging/segmentation/picai/runner.py [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+    lgr.info("Fuse Eval", {"attrs": ["bold", "underline"]})
+
+    # create evaluator
+    evaluator = EvaluatorDefault()
+
+    # run
+    results = evaluator.eval(
+        ids=None,
+        data=os.path.join(paths["inference_dir"], infer["infer_filename"]),
+        metrics={},
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/runner.py:328]: for now the loss is dice loss which is 1-dice.
the only metric I know for current semantic segmentation is dice.
maybe I will call the dice metric, will also ask Moshiko if it can return just the loss somehow

File: fuse_examples/imaging/segmentation/picai/runner.py [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+
+    # train
+    if "train" in cfg["run.running_modes"]:
+        run_train(NDict(cfg["paths"]), NDict(cfg["train"]))
+
+    # infer
+    if "infer" in cfg["run.running_modes"]:
+        run_infer(NDict(cfg["infer"]), NDict(cfg["paths"]), NDict(cfg["train"]))
+    #
+    # analyze - skipping as it crushes without metrics
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/runner.py:351]: same as above, will try to convert to either of the 2 solutions.

File: fuse_examples/tests/conf/config.yaml (hunk 1 above)
Comment (itaijj) [reply fuse_examples/tests/conf/config.yaml:30]: I actually override it in the test,
not sure what is the better way to do it,
@mosheraboh any idea?

File: fuse_examples/tests/test_segmentation_picai.py (hunk 2 above)
Comment (itaijj) [reply fuse_examples/tests/test_segmentation_picai.py:49]: same as above

File: fuseimg/datasets/picai.py (hunk 3 above)
Comment (itaijj) [reply fuseimg/datasets/picai.py:74]: what do you mean by clearer?
@mosheraboh any idea?

File: fuseimg/datasets/picai.py (hunk 4 above)
Comment (itaijj) [reply fuseimg/datasets/picai.py:86]: done

File: fuse_examples/imaging/segmentation/picai/conf/config_template.yaml [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  cache_dir : ${paths.working_dir}/cache/PICAI
+  clinical_file : ${oc.env:PICAI_DATA_PATH}/labels/picai_labels-main/clinical_information/marksheet.csv
+  data_misc_dir : ${paths.working_dir}/data_misc
+  data_split_filename:  split.pkl
+  data_dir :  ${oc.env:PICAI_DATA_PATH}/images
+  seg_dir : ${oc.env:PICAI_DATA_PATH}/labels/picai_labels-main/csPCa_lesion_delineations/human_expert/resampled
+run: 
+  running_modes : ['train', 'infer', 'eval']
+train:
+  target : seg3d
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/conf/config_template.yaml:15]: will add comments in hydra and readme once we finalize the code

File: fuse_examples/tests/test_segmentation_picai.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    from fuse_examples.imaging.segmentation.picai.runner import run_train, run_infer, run_eval
+
+
+def run_picai(root: str) -> None:
+    initialize(config_path="../imaging/segmentation/picai/conf", job_name="test_app")
+    cfg = compose(config_name="config_template")
+    print(str(cfg))
+
+    cfg = NDict(hydra.utils.instantiate(cfg))
+    cfg["paths.working_dir"] = root
```
Comment (SagiPolaczek) [inline fuse_examples/tests/test_segmentation_picai.py:47]: I think that you change your `paths.working_dir` after you instantiate the config. Thus the cache dir doesn't have the expected value: `${paths.working_dir}/cache/PICAI`.

Please double check me on that :)

Comment (itaijj) [reply fuse_examples/tests/test_segmentation_picai.py]: Thanks, fixed the overide to make it work properly

File: fuse_examples/imaging/segmentation/picai/conf/config_template.yaml [hunk 12]
```diff
@@ -0,0 +1,53 @@
+paths:
+  working_dir : .
+  model_dir : ${paths.working_dir}/model_dir
+  inference_dir : ${paths.working_dir}/model_new/infer_dir
+  eval_dir : ${paths.working_dir}/model_new/eval_dir
+  cache_dir : ${paths.working_dir}/cache/PICAI
+  clinical_file : ${oc.env:PICAI_DATA_PATH}/labels/picai_labels-main/clinical_information/marksheet.csv
```
Comment (mosheraboh) [inline fuse_examples/imaging/segmentation/picai/conf/config_template.yaml:7]: Did you add PICAI_DATA_PATH to env.sh file on CCC?

File: fuse_examples/imaging/segmentation/picai/conf/config_template.yaml [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+  reset_cache: False
+  num_workers : 16
+  num_folds : 5
+  train_folds : [0,1,2,3]
+  validation_folds : [4]
+  batch_size: 16
+  learning_rate : 1e-3
+  weight_decay :  0
+  run_sample : 0 #if 0 - used all samples otherwise sample only run_sample samples ( for test purpouse )
+  resume_checkpoint_filename :
```
Comment (mosheraboh) [inline fuse_examples/imaging/segmentation/picai/conf/config_template.yaml:27]: set to null explicitly

File: fuse_examples/imaging/segmentation/picai/conf/config_template.yaml [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+  batch_size: 16
+  learning_rate : 1e-3
+  weight_decay :  0
+  run_sample : 0 #if 0 - used all samples otherwise sample only run_sample samples ( for test purpouse )
+  resume_checkpoint_filename : 
+  trainer:
+    accelerator : gpu
+    devices : 1
+    num_epochs : 100
+    ckpt_path :
```
Comment (mosheraboh) [inline fuse_examples/imaging/segmentation/picai/conf/config_template.yaml:32]: set to null explicitly

File: fuse_examples/imaging/segmentation/picai/runner.py [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+
+
+#################################
+# Train Template
+#################################
+def run_train(paths: NDict, train: NDict) -> None:
+    # ==============================================================================
+    # Logger
+    # ==============================================================================
+    # fuse_logger_start(output_path=paths["model_dir"], console_verbose_level=logging.INFO)
```
Comment (mosheraboh) [inline fuse_examples/imaging/segmentation/picai/runner.py:91]: You can remove the fuse_logger_start and move to standard prints. 
We will deprecate it soon.

File: fuse_examples/imaging/segmentation/picai/runner.py [hunk 16]
```diff
@@ -0,0 +1,381 @@
...
+    lgr.info("Model: Done", {"attrs": "bold"})
+
+    lgr.info("\nFuse Train", {"attrs": ["bold", "underline"]})
+
+    lgr.info(f'model_dir={paths["model_dir"]}', {"color": "magenta"})
+    lgr.info(f'cache_dir={paths["cache_dir"]}', {"color": "magenta"})
+
+    #### Train Data
+    # split to folds randomly - temp
+    dataset_all = PICAI.dataset(
```
Comment (mosheraboh) [inline fuse_examples/imaging/segmentation/picai/runner.py:110]: Our current recommendation is to move to lightning data module to support multigpu. 
(but it's not a must)

File: fuse_examples/imaging/segmentation/picai/runner.py [hunk 17]
```diff
@@ -0,0 +1,381 @@
...
+        train=True,
+    )
+
+    validation_dataset = PICAI.dataset(paths=paths, cfg=train, reset_cache=False, sample_ids=validation_sample_ids)
+
+    ## Create sampler
+    lgr.info("- Create sampler:")
+    sampler = BatchSamplerDefault(
+        dataset=train_dataset,
+        balanced_class_name="data.gt.classification",  # gt_label, TODO - make diff label for balance-sampler
```
Comment (mosheraboh) [inline fuse_examples/imaging/segmentation/picai/runner.py:147]: What is the todo comment?

File: fuse_examples/imaging/segmentation/picai/runner.py [hunk 18]
```diff
@@ -0,0 +1,381 @@
...
+
+    # ====================================================================================
+    #  Loss and metrics
+    # ====================================================================================
+    # TODO - add a classification loss - add head to the bottom of the unet
+    losses = {}
+    losses["segmentation"] = LossDefault(
+        pred="model.seg",
+        target="data.gt.seg",
+        callable=monai.losses.DiceLoss(to_onehot_y=True, softmax=False),
```
Comment (mosheraboh) [inline fuse_examples/imaging/segmentation/picai/runner.py:194]: do we have diceloss in fuse?

File: fuse_examples/imaging/segmentation/picai/runner.py [hunk 19]
```diff
@@ -0,0 +1,381 @@
...
+    evaluator = EvaluatorDefault()
+    metrics = OrderedDict(
+        [
+            ("dice", MetricDice(pred="model.seg", target="data.gt.seg")),
+        ]
+    )
+    # define iterator
+
+    def data_iter() -> NDict:
+        # set seed
```
Comment (mosheraboh) [inline fuse_examples/imaging/segmentation/picai/runner.py:334]: remove comment

File: fuse_examples/imaging/segmentation/picai/runner.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mosheraboh) [reply fuse_examples/imaging/segmentation/picai/runner.py:375]: You have metric now, so you can delete this comment.

File: fuseimg/datasets/picai.py [hunk 20]
```diff
@@ -0,0 +1,350 @@
...
+    def dataset(
+        paths: NDict,
+        cfg: NDict,
+        reset_cache: bool = True,
+        sample_ids: Optional[Sequence[Hashable]] = None,
+        train: bool = False,
+        run_sample: int = 0,
+    ) -> DatasetDefault:
+        """
+        Creates Fuse Dataset single object (either for training, validation and test or user defined set)
```
Comment (mosheraboh) [inline fuseimg/datasets/picai.py:300]: update the comments in here

File: fuse_examples/imaging/segmentation/picai/conf/config_template.yaml (hunk 12 above)
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/conf/config_template.yaml:7]: yes, otherwise the test would not run...

File: fuse_examples/imaging/segmentation/picai/conf/config_template.yaml (hunk 13 above)
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/conf/config_template.yaml:27]: ok

File: fuse_examples/imaging/segmentation/picai/conf/config_template.yaml (hunk 14 above)
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/conf/config_template.yaml:32]: done

File: fuse_examples/imaging/segmentation/picai/runner.py (hunk 15 above)
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/runner.py:91]: ok

File: fuse_examples/imaging/segmentation/picai/runner.py (hunk 17 above)
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/runner.py:147]: I'l remove it

File: fuse_examples/imaging/segmentation/picai/runner.py (hunk 18 above)
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/runner.py:194]: we have something, but it's not suited.
will create our fixed version after finishing current code PR fix

File: fuseimg/datasets/picai.py (hunk 20 above)
Comment (itaijj) [reply fuseimg/datasets/picai.py:300]: done

File: fuse_examples/imaging/segmentation/picai/runner.py (hunk 16 above)
Comment (itaijj) [reply fuse_examples/imaging/segmentation/picai/runner.py:110]: if there will be a future requirement for such multigpu training on this example I'l change it

**Review by mosheraboh — APPROVED** [review]
